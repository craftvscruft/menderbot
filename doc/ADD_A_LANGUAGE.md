# Add a new language

See this [GitHub issue](https://github.com/craftvscruft/menderbot/issues/5) for the current status of language support.

## Generate parser

Update `scripts/generate_parsers.sh` in the "Generate parsers" section to add your new language. It's expecting directories similar to `vendor/grammars-v4`.

```sh
PYTHON_GRAMMAR_DIR="$STARTING_DIR/vendor/grammars-v4/python/python"
gen_parser Python \
    # Grammar dir
    "$PYTHON_GRAMMAR_DIR" \
    # Work dir:
    "$PYTHON_GRAMMAR_DIR/Python3"
```

* One or more .g4 grammar files in the grammar directory
* A `transformGrammar.py` script in the work directory
* A `*Base.py` files in the work directory

If successful, it will create new parser code in `menderbot/antlr_generated` with the others.

## Implement a LanguageStrategy

Start in `menderbot/code.py`. This pattern is rapidly changing, more coming here when it settles down. 

Please [reach out](https://github.com/craftvscruft/menderbot/issues/5) if you are interested in adding a target language.

## Test the plumbing

Add an example to `scripts/smoke.sh` and run it.