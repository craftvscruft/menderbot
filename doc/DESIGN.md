# Design

Menderbot was created to demonstrate the principles of the [Mechanized Mending Manifesto](https://mender.ai/docs/intro), summarized:

* The dominant challenge in software development is maintenance
* LLMs can help
* Humans, conventional algorithms, and LLMs have different strengths
  * They do not replace each other
  * Effective systems will carefully combine them

Following these principles, Menderbot aims to:

* Optimize for working on large existing codebases
* Treat LLM output as useful but untrustworthy
* Default to safety
* Play well with other automation

## Safeguards

* LLM output is always labeled
* Avoid using LLM output in executable code
* Require explicit consent to use LLM on repo
* **pending**: Price confirmation for batch operations

### LLM consent

Assume the user has multiple API keys and only one may be authorized.

### LLM output in code

Since LLM output is untrustworthy, we avoid using it to modify code where possible. Here are some safer uses.

* Comments, sanitized to prevent code injection
* Type annotations, sanitized and type-checked
* Refactoring suggestions executed by deterministic syntax-tree operations

## Language-aware

We plan to support these language families at minimum, with syntax awareness.

* JavaScript / TypeScript
* C / C++
* Java
* Python

## Technical decisions

### Python

Python is currently the defacto language for AI development, so it's a reasonable pick for an experimental AI project. However, other languages such as Go and Rust could make the distribution of a command-line tool easier.

### Antlr 
We use the [Antlr Python Target](https://github.com/antlr/antlr4/blob/master/doc/python-target.md) to parse source code so that we can make updates in consistent syntax-aware way.

Pros:
* Supports a wide variety of languages with community maintained grammars available
* Mature and widely adopted
* Allows us to stay in pure Python both at build and runtime

Cons:
* Not Python native, API very hairy to use compared to the 
* Requires running Java to re-generate code on grammar changes

Unknowns:
* Can we parse syntax errors and keep going in "the right way"?

#### tree-sitter (Runner Up)

Pros:
* Supports a wide variety of languages with community maintained grammars available
* Can parse files with syntax errors and keep going

Cons:
* Requires shipping with a platform-specific binary built from C

Other alternatives considered: 

Lark, LibCST (parsing Python)

