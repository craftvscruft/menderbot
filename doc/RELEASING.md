# Releasing

If you are a member of the Craft vs Cruft org, you can release Menderbot using git [tags](https://git-scm.com/book/en/v2/Git-Basics-Tagging).

```sh
. ./scripts/version.sh
```

Commit everthing. Do a dry run of the version bump.

```sh
./venv/bin/bumpversion --current-version $MENDERBOT_VERSION patch --dry-run --verbose
```

Version bump:

```sh
./venv/bin/bumpversion --current-version $MENDERBOT_VERSION patch
```

```sh
gh auth login

# For a prerelease:
gh release create --prerelease "v${MENDERBOT_VERSION}"

# For a real release:
gh release create "v$MENDERBOT_VERSION"
```

Hit enter a few times, leave a note. Check the workflows for progress. 