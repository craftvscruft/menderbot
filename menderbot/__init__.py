__version__ = "0.0.3-pre"  # Update README badge for new versions.

from menderbot.__main__ import cli
