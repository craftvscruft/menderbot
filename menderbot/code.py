def line_indent(line):
    count = len(line)-len(line.lstrip())
    return line[:count]

def function_indent(code):
    second_line_start = code.find('\n') + 1
    no_first_line = code[second_line_start:]
    if no_first_line.find('\n') > -1:
        second_line_end = second_line_start + no_first_line.find('\n')
        second_line = code[second_line_start:second_line_end]
    else:
        second_line = no_first_line
    return line_indent(second_line)

def reindent(text, indent):
    lines = text.split('\n')
    indented_lines = [indent + line.lstrip() for line in lines]
    return '\n'.join(indented_lines)