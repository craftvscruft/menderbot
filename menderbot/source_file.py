import tempfile
from typing import Iterable
from dataclasses import dataclass
import os
import itertools
from pathlib import Path
from charset_normalizer import from_path


@dataclass
class Insertion:
    text: str
    line_number: int
    label: str


def insert_in_lines(lines: Iterable[str], insertions: Iterable[Insertion]):
    lines = iter(lines)
    last_line = 1
    for insertion in insertions:
        for line in itertools.islice(lines, insertion.line_number - last_line):
            yield line
            last_line += 1
        yield insertion.text
        yield "\n"
    yield from lines


class SourceFile:
    def __init__(self, path: str):
        self.path = path
        self.encoding = None
        self._initial_modified_time = os.path.getmtime(path)

    def load_source_as_utf8(self):
        loaded = from_path(self.path)
        best_guess = loaded.best()
        self.encoding = best_guess.encoding
        return best_guess.output(encoding="utf_8")

    def is_unicode(self):
        return self.encoding.startswith("utf")

    def update_file(self, insertions: Iterable[Insertion], suffix):
        path_obj = Path(self.path)
        with path_obj.open("r", encoding=self.encoding) as filehandle:
            if self.modified_after_loaded():
                raise Exception(
                    f"File '{self.path}' was externally modified, try again."
                )
            new_lines = list(insert_in_lines(lines=filehandle, insertions=insertions))
            out_file = path_obj.with_suffix(f"{path_obj.suffix}{suffix}")
            self._write_result(new_lines, out_file)

    def _write_result(self, lines, output_file: Path):
        with tempfile.TemporaryDirectory() as tempdir:
            my_tempfile: Path = Path(tempdir) / "output.txt"
            with my_tempfile.open("w") as filehandle:
                for line in lines:
                    filehandle.write(line)
            my_tempfile.replace(output_file)

    def modified_after_loaded(self):
        return os.path.getmtime(self.path) > self._initial_modified_time
