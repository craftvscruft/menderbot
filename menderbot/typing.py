import logging
import os
import re

from menderbot import python_cst
from menderbot.source_file import Insertion, SourceFile

logger = logging.getLogger("typing")


def process_untyped_functions(source_file: SourceFile):
    path = source_file.path
    logger.info('Processing "%s"...', path)
    _, file_extension = os.path.splitext(path)
    if not file_extension == ".py":
        logger.info('"%s" is not a Python file, skipping.', path)
        return
    source = source_file.load_source_as_utf8()

    for fn_ast in python_cst.collect_function_asts(source):
        yield (fn_ast, what_needs_typing(fn_ast))


def parse_type_hint_answer(text: str) -> list:
    def line_to_tuple(line: str) -> tuple:
        [ident, new_type] = line.split(":")
        new_type = re.sub(r"\bList\b", "list", new_type)
        # NoReturn has valid uses but is unlikely, assume we mean None
        new_type = re.sub(r"\bNoReturn\b", "None", new_type)
        return (ident.strip(), new_type.strip())

    lines = text.strip().splitlines()
    hints = [line_to_tuple(line) for line in lines if ":" in line]
    return [hint for hint in hints if hint[0] != "self" and hint[1].lower() != "any"]


def add_type_hints(
    function_ast: python_cst.AstNode,
    hints: list[tuple[str, str]],
    imports: list[tuple[str, str]],
) -> list:
    print("function_node", function_ast.props)
    function_name = function_ast.props["name"]
    sig_ast = function_ast.children_filtered(kind=python_cst.KIND_SIGNATURE)[0]
    def_param_nodes = sig_ast.children_filtered(kind=python_cst.KIND_PARAM)
    return_type = function_ast.props.get(python_cst.PROP_RETURN_TYPE)
    insertions = []
    common_typing_import_names = ["Optional", "Callable", "NamedTuple", "Any", "Type"]

    def add_needed_imports(new_type: str):
        new_type_symbols = re.findall(r"\b\w+\b", new_type)
        for new_type_symbol in new_type_symbols:
            if (
                new_type_symbol in common_typing_import_names
                and ("typing", new_type_symbol) not in imports
            ):
                imports.append(("typing", new_type_symbol))
                insertions.append(
                    Insertion(
                        text=f"from typing import {new_type_symbol}",
                        line_number=1,
                        label="type_import",
                    )
                )

    for ident, new_type in hints:
        add_needed_imports(new_type)
        for param_node in def_param_nodes:
            if param_node.props.get("name") == ident:
                insertions.append(
                    Insertion(
                        text=f": {new_type}",
                        line_number=param_node.src_range.end.line,
                        col=param_node.src_range.end.col,
                        inline=True,
                        label=function_name,
                    )
                )
        if ident == "return" and not return_type:
            signature_ast = function_ast.children_filtered(
                kind=python_cst.KIND_SIGNATURE
            )[0]
            insertions.append(
                Insertion(
                    text=f" -> {new_type}",
                    line_number=signature_ast.src_range.end.line,
                    col=signature_ast.src_range.end.col,
                    inline=True,
                    label=function_name,
                )
            )

    return insertions


def what_needs_typing(fn_ast: python_cst.AstNode) -> list[str]:
    name = fn_ast.props["name"]
    sig_ast = fn_ast.children_filtered(kind=python_cst.KIND_SIGNATURE)[0]
    param_asts = sig_ast.children_filtered(kind=python_cst.KIND_PARAM)
    return_type = fn_ast.props.get(python_cst.PROP_RETURN_TYPE)
    needs_typing = [
        param_ast.props["name"]
        for param_ast in param_asts
        if "type" not in param_ast.props
        and param_ast.props["name"] not in ["self", "cls"]
    ]
    if not return_type and name != "__init__":
        needs_typing.append("return")
    return needs_typing


# def get_function_param_nodes(
#     node: PythonParser.FuncdefContext,
# ) -> Generator[PythonParser.Def_parameterContext, None, None]:
#     args_list_node: PythonParser.TypedargslistContext = node.typedargslist()
#     # args_node : Optional[PythonParser.ArgsContext] = params_node.args()
#     # def_params_node: Optional[PythonParser.Def_parametersContext] = params_node.def_parameters()
#     # kwargs_node: Optional[PythonParser.KwargsContext] = params_node.kwargs()
#     if args_list_node:
#         def_params_nodes: list[
#             PythonParser.Def_parameterContext
#         ] = args_list_node.def_parameters()
#         for def_params_node_untyped in def_params_nodes:
#             def_params_node: PythonParser.Def_parametersContext = def_params_node_untyped
#             for def_param_node_untyped in def_params_node.def_parameter():
#                 def_param_node: PythonParser.Def_parameterContext = def_param_node_untyped
#                 yield def_param_node
