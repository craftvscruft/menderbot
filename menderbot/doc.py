import logging
import os
from typing import Callable

from menderbot.code import LANGUAGE_STRATEGIES, node_start_line, node_str
from menderbot.source_file import Insertion, SourceFile

logger = logging.getLogger("doc")


def init_logging() -> None:
    """
    Initializes the logging module.
    """
    logger.setLevel(logging.DEBUG)
    # create console handler with a higher log level
    stream_handler = logging.StreamHandler()
    stream_handler.setLevel(logging.INFO)
    logger.addHandler(stream_handler)


init_logging()


def document_file(source_file: SourceFile, doc_gen: Callable) -> list[Insertion]:
    """
    Generates documentation for functions in the sourcefile that don't have it
    using the supplied `doc_gen` callable.

    Return a list of insertions which the caller can use to update the file.

    If the file extension has no language strategy, the function logs a message and returns an empty list.
    """
    path = source_file.path
    logger.info('Processing "%s"...', path)
    _, file_extension = os.path.splitext(path)
    language_strategy = LANGUAGE_STRATEGIES.get(file_extension)
    if not language_strategy:
        logger.info('Unrecognized extension "%s", skipping.', file_extension)
        return []

    source = source_file.load_source_as_utf8()
    tree = language_strategy.parse_source_to_tree(source)
    insertions = []
    for node in language_strategy.get_function_nodes(tree):
        if not language_strategy.function_has_comment(node):
            name = language_strategy.get_function_node_name(node)
            logger.info('Found undocumented function "%s"', name)
            code = node_str(node)
            comment = doc_gen(code, file_extension)
            function_start_line = node_start_line(node)
            doc_start_line = (
                function_start_line + language_strategy.function_doc_line_offset
            )
            if comment:
                logger.info("Documenting with: %s", comment)
                logger.info("For code: %s", code)
                insertions.append(
                    Insertion(text=comment, line_number=doc_start_line, label=name)
                )
    return insertions
