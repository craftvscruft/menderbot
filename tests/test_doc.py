import os
from typing import Iterable

from approvaltests.approvals import verify

from menderbot.code import LANGUAGE_STRATEGIES, function_indent, reindent
from menderbot.doc import document_file
from menderbot.source_file import Insertion, SourceFile, insert_in_lines


class FakeSourceFile(SourceFile):
    def __init__(self, path: str, text: str):
        self.path = path
        self.encoding = None
        self._initial_modified_time = None
        self.text = text
        self.original_text = text

    def update_file(self, insertions: Iterable[Insertion], suffix: str) -> None:
        del suffix
        if len(insertions) == 0:
            return
        # Split keeping delimiter, so each line will have a "\n"
        old_lines = self.text.splitlines(True)
        new_lines = list(insert_in_lines(lines=old_lines, insertions=insertions))
        self.text = "".join(new_lines)

    def load_source_as_utf8(self):
        return bytes(self.text, "utf-8")

    def is_unicode(self):
        return True

    def modified_after_loaded(self):
        return False


def test_parse_with_fake_source_file():
    source_file = FakeSourceFile(
        "my/path.py",
        """
def foo():
    print('Hello World')
    
def foo():
    \"\"\"Hello\"\"\"
    print('Hello World')
""",
    )

    path = source_file.path
    _, file_extension = os.path.splitext(path)
    language_strategy = LANGUAGE_STRATEGIES.get(file_extension)
    assert language_strategy

    source = source_file.load_source_as_utf8()
    tree = language_strategy.parse_source_to_tree(source)
    function_nodes = language_strategy.get_function_nodes(tree)
    assert len(function_nodes) == 2
    assert not language_strategy.function_has_comment(function_nodes[0])
    assert language_strategy.function_has_comment(function_nodes[1])


def test_do_nothing_for_unknown_extension():
    def generate_fake_doc(code, file_extension):
        del code
        del file_extension
        raise AssertionError("Should not be called")

    source_file = FakeSourceFile(
        "my/path.unknown",
        """
def foo():
    print('Hello World')
""",
    )

    insertions = document_file(source_file, generate_fake_doc)
    source_file.update_file(insertions, suffix="")
    assert source_file.text == source_file.original_text


def test_adds_python_doc():
    def generate_fake_doc(code, file_extension):
        assert file_extension == ".py"
        indent = function_indent(code)
        return reindent('"""Hi!"""', indent)

    source_file = FakeSourceFile(
        "my/path.py",
        """
def foo():
    print('Hello World')
""",
    )

    insertions = document_file(source_file, generate_fake_doc)
    source_file.update_file(insertions, suffix="")
    verify(source_file.text)


def test_passes_in_code_to_generator():
    code_received = None

    def generate_fake_doc(code, file_extension):
        nonlocal code_received
        assert file_extension == ".py"
        code_received = code
        return '  """Hi!"""'

    source_file = FakeSourceFile(
        "my/path.py",
        """
def foo():
    print('Hello World')
""",
    )

    document_file(source_file, generate_fake_doc)
    assert code_received == "def foo():\n    print('Hello World')\n\n"
