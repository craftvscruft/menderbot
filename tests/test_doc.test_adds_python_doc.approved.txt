
def foo():
    """Hi!"""
    print('Hello World')
